Integrate resource monitor

t-davidbrown-msft opened the pull request:
- add concurrency tuner to internal TransferManager constructor
- add _concurrencyTuner instance field
- add StartConcurrncyTuner private method to TransferManager, which passes cancellation token to the dependencies
- add StopConcurrencyTunerWhenTransfersComplete, which looks at the states for all jobs, then runs _cancellationTokenSource.Cancel() when all states are complete
- nested ResourceMonitor in ConcurrencyTuner
- create region for worker helper methods
- replace the parameter in the Worker method with an instance variable
- refactored Worker method
- created cancellationToken field
- delete CancellationTokenSource
- remove StopMonitoring since this is now controlled by CancellationTokenSource in the TransferManager
- Added `MinimumChunkReaders` and `MaxJobReaders` to the DataMovementConstants class
- add default value for TransferErrorMode to TransferManagerOptions
- add default value for MaximumConcurrency to TransferManagerOptions
- add MaximumCpuUsage property  with default value
- add MonitoringInterval property

Commit:
create cpu monitor

- create method for monitoring CPU usage that works for all frameworks using System.Diagnostics.Process
- create one test for ensuring that the monitor works

Commit:
Merge branch 'main' into cpu-monitor
Merging the latest changes into my feature branch

Commit:
update cpu monitor logic

- add CpuMonitor interface
- add CpuMonitor tests
- change CurrentProcessorTime to be the current time on the processor minute the previous

Commit:
finished baseline cpu tuner

- removed unnecessary refactor
- add SimulateCpuLoad to test suite to simulate loading the CPU
- add CoreCount Property to replace function calls to Environment.ProcessorCount

Commit:
add testing

- readd default constructor
- add tests for monitoring interval that is below 100 milliseconds

Commit:
create memory monitor

- create processor directive to create memory monitor for netstandard2.0 or greater, but this is not a great implementation
- create method for MonitorMemoryUsage is otherwise good
- create tests to check whether it is below 1, above 0, and not null

Commit:
add memory check for netstandard20

Commit:
finish cpu and memory monitors

- add cpu monitor returns a float between 0 and 1
- add memory monitor currently return a double between 0 and 1 (this is the current amount of memory a process is using divided by the total amount of memory on the host)
- add tests for cpu monitor and memory monitor. Tests check for 0, less than 1, and null

Commit:
Merge branch 'main' into resource-moni o
merge main into feature branch

Commit:
rename CpuMonitor to ResourceMonitor

Commit:
fix merge issue and add changes

- remove interface
- change private properties to fields
- group private fields and properties
- remove public default constrcutor
- combine CPU and memory into one task and run on single thread
- switch order in resource monitor: take reading first, then sleep
- memory monitor returns Bytes instead of percentage
- remove all functionality for getting total memory

Commit:
change interface to concrete class

Commit:
update tests

Commit:
integrated resource monitor; prep for testing

- add concurrency tuner to internal TransferManager constructor
- add _concurrencyTuner instance field
- add StartConcurrncyTuner private method to TransferManager, which passes cancellation token to the dependencies
- add StopConcurrencyTunerWhenTransfersComplete, which looks at the states for all jobs, then runs _cancellationTokenSource.Cancel() when all states are complete
- nested ResourceMonitor in ConcurrencyTuner
- create region for worker helper methods
- replace the parameter in the Worker method with an instance variable
- refactored Worker method
- created cancellationToken field
- delete CancellationTokenSource
- remove StopMonitoring since this is now controlled by CancellationTokenSource in the TransferManager
- Added MinimumChunkReaders and MaxJobReaders to the DataMovementConstants class
- add default value for TransferErrorMode to TransferManagerOptions
- add default value for MaximumConcurrency to TransferManagerOptions
- add MaximumCpuUsage property  with default value
- add MonitoringInterval property

Commit:
integrated resource monitor; prep for testing

- add concurrency tuner to internal TransferManager constructor
- add _concurrencyTuner instance field
- add StartConcurrncyTuner private method to TransferManager, which passes cancellation token to the dependencies
- add StopConcurrencyTunerWhenTransfersComplete, which looks at the states for all jobs, then runs _cancellationTokenSource.Cancel() when all states are complete
- nested ResourceMonitor in ConcurrencyTuner
- create region for worker helper methods
- replace the parameter in the Worker method with an instance variable
- refactored Worker method
- created cancellationToken field
- delete CancellationTokenSource
- remove StopMonitoring since this is now controlled by CancellationTokenSource in the TransferManager
- Added MinimumChunkReaders and MaxJobReaders to the DataMovementConstants class
- add default value for TransferErrorMode to TransferManagerOptions
- add default value for MaximumConcurrency to TransferManagerOptions
- add MaximumCpuUsage property  with default value
- add MonitoringInterval property

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.DataMovement](https://apiview.dev/Assemblies/Review/fffb907bd4224ac4b0f95dfb164a72e3?revisionId=24e52c6e41dd4576aa9f5be883219e98)

Commit:
add default argument to tests

Commit:
add default argument to tests

Commit:
remove async/await so resource monitor runs in background

Commit:
remove async/await so resource monitor runs in background

Commit:
create concurrency tuner tests

Commit:
create concurrency tuner tests

Commit:
fix failing tests

Commit:
fix failing tests

Commit:
Merge branch 'main' into integrate-resource-monitor

Commit:
Merge branch 'main' into resource-monitor

Commit:
remove ConcurrencyTuner from TransferManager API

- remove ConcurrencyTuner from TransferManager API
- add chunkProcessor to ConcurrencyTuner API

Commit:
add EnableConcurrencyTuner bool

Commit:
add ConcurrencyTunerEnabled option

add ConcurrencyTunerEnabled to TransferManagerOptions
set default for concurrencyTunerEnabled add bool to transfer manager
add checks for concurrencyTunerEnabled to create and start
concurrencyTuner
set ConcurrencyTunerEnabled to true

Commit:
add xml comment to ConncurrencyTunerEnabled

Commit:
fixed merge from remote
